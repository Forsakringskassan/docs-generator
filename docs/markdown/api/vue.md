---
title: Documenting Vue components
short-title: Vue Components
layout: article
---

API documentation for Vue components can be enabled using the {@link vueFileReader}:

```ts
import {
    Generator,
    frontMatterFileReader,
    vueFileReader,
} from "@forsakringskassan/docs-generator";

const docs = new Generator({
    site: { name: "My Awesome Site" },
    setupPath: "setup.ts",
});

/* --- cut above --- */

await docs.build([
    {
        include: ["./docs/**/*.md"],
        fileReader: frontMatterFileReader,
    },
    {
        include: ["./src/**/*.vue"],
        fileReader: vueFileReader,
    },
]);

/* --- cut below --- */

export {}; // for TLA to work
```

This will read each component and generate API documentation which can later be included in markdown with:

```md
:::api
vue:AwesomeComponent
:::
```

For details about how to document Vue components see [vue-docgen-api documentation](https://vue-styleguidist.github.io/docs/Documenting.html).

## Example

:::api
vue:VueExample
:::
