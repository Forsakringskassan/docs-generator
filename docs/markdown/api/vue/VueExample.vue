<script setup lang="ts">
defineProps({
    /**
     * Lorem ipsum dolor
     * sit amet.
     *
     * Must be one of:
     *
     * - `"a"`
     * - `"b"`
     * - `"c"`
     */
    foo: {
        type: String,
        required: false,
        default: "foobar",
    },
    /**
     * Lorem ipsum dolor sit amet, consectetur adipiscing elit. Morbi eu mollis
     * leo. Fusce eget ullamcorper augue. Aenean sodales fringilla purus, ac
     * euismod massa semper sed.
     */
    bar: {
        type: String,
        required: true,
    },
    /**
     * Lorem ipsum dolor sit amet.
     *
     * @deprecated Replaced by something else.
     */
    obsolete: {
        type: Number,
        required: false,
        default: 42,
    },
});
defineEmits<{
    /**
     * Lorem ipsum
     * dolor sit amet
     *
     * @param id - Lorem ipsum
     * @param value - Lorem ipsum
     */
    change: [id: number, value: string];
    /**
     * Lorem ipsum
     * dolor sit amet
     *
     * @param value - Lorem ipsum
     */
    update: [value: string];
    /**
     * Lorem ipsum dolor sit amet
     *
     * @deprecated Will be gone later.
     * @param value - Lorem ipsum
     */
    obsolete: [value: string];
}>();
</script>

<template>
    <!-- @slot Lorem ipsum dolor sit amet -->
    <slot name="default"></slot>

    <!--
         @slot Lorem ipsum dolor sit amet
         @binding {number} foo Lorem ipsum dolor sit amet
         @binding {number} bar Lorem ipsum dolor sit amet
    -->
    <slot name="foo" v-bind="{ foo: 1, bar: 2, baz: 'spam' }"></slot>

    <!--
         @slot Lorem ipsum dolor sit amet.
         @deprecated Use another slot instead.
    -->
    <slot name="obsolete"></slot>
</template>
