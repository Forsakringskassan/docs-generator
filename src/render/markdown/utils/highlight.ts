import hljs from "highlight.js/lib/common";

const HTML_ESCAPE_REPLACE_RE = /[&<>"]/g;
const HTML_REPLACEMENTS: Record<string, string> = {
    "&": "&amp;",
    "<": "&lt;",
    ">": "&gt;",
    '"': "&quot;",
};

function escapeHtml(str: string): string {
    return str.replace(HTML_ESCAPE_REPLACE_RE, (ch) => HTML_REPLACEMENTS[ch]);
}

function wrap(content: string, language: string): string {
    const markup = /* HTML */ `
        <code class="hljs lang-${language}" tabindex="0">${content}</code>
    `;
    return markup.trim();
}

/**
 * @internal
 */
export function highlight({
    source,
    language,
}: {
    source: string;
    language: string;
}): string {
    switch (language) {
        case "":
        case "import":
        case "mermaid":
        case "plaintext":
            return `<code>${escapeHtml(source)}</code>`;
        case "vue":
            return highlight({ source, language: "html" });
        case "html":
    }
    try {
        const { value } = hljs.highlight(source, { language });
        return wrap(value, language);
    } catch (err) {
        return wrap(String(err), language);
    }
}
