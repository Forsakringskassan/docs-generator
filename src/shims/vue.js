/* eslint-env browser */
export default window.Vue;
export const defineComponent = window.Vue.defineComponent;
export const toDisplayString = window.Vue.toDisplayString;
export const openBlock = window.Vue.openBlock;
export const createElementVNode = window.Vue.createElementVNode;
export const createElementBlock = window.Vue.createElementBlock;
export const createTextVNode = window.Vue.createTextVNode;
export const resolveComponent = window.Vue.resolveComponent;
export const withCtx = window.Vue.withCtx;
export const createBlock = window.Vue.createBlock;
