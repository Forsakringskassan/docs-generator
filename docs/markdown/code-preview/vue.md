---
title: Vue examples
layout: article
---

## Options API

```import
vue-options.vue
```

## Composition API

```import
vue-composition.vue
```

## Snippets

The source code may also contains snippets which are cut from the rendered source code.

This is mostly useful in combination with extracting the example source code and running external tools, e.g. compiling the example might require additional imports or declarations.

**Input:**

````md
```vue
<script setup>
import { ref } from "vue";

const name = "World";
</script>

<template>
    <!-- cut above -->
    <p>Hello {{ name }}!</p>
    <!-- cut below -->
</template>
```
````

**Output:**

```vue
<script setup>
const name = "World";
</script>

<template>
    <!-- cut above -->
    <p>Hello {{ name }}!</p>
    <!-- cut below -->
</template>
```

The following instructions are available:

- `<!-- cut above -->` removes all lines above the instruction.
- `<!-- cut below -->` removes all lines below the instruction.
- `<!-- cut begin -->` and `<!-- cut end -->` removes all lines between the two instructions.

Blank lines above or below the instructions will always be trimmed as well.
