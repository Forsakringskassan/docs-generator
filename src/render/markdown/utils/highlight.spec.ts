import { highlight } from "./highlight";

it("should render syntax highlighting", () => {
    expect.assertions(1);
    const source = `export const meaningOfLife = 42;`;
    expect(highlight({ source, language: "ts" })).toMatchSnapshot();
});

it("should render vue as html", () => {
    expect.assertions(1);
    const source = /* HTML */ `
        <template>
            <p>lorem ipsum dolor sit amet</p>
        </template>

        <script>
            import { defineComponent } from "vue";
            export default defineComponent({
                name: "MyAwesomeComponent",
            });
        </script>
    `;
    expect(highlight({ source, language: "vue" })).toMatchSnapshot();
});

it("should render plaintext without syntax highlighing", () => {
    expect.assertions(1);
    const source = `export const meaningOfLife = 42;`;
    expect(highlight({ source, language: "plaintext" })).toMatchSnapshot();
});

it("should handle missing language as plaintext", () => {
    expect.assertions(1);
    const source = `export const meaningOfLife = 42;`;
    expect(highlight({ source, language: "" })).toMatchSnapshot();
});

it("should escape html in plaintext", () => {
    expect.assertions(1);
    const source = `<script>alert("xss")</script>`;
    expect(highlight({ source, language: "plaintext" })).toMatchSnapshot();
});
