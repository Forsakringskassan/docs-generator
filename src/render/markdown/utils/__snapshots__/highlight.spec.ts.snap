// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`should escape html in plaintext 1`] = `"<code>&lt;script&gt;alert(&quot;xss&quot;)&lt;/script&gt;</code>"`;

exports[`should handle missing language as plaintext 1`] = `"<code>export const meaningOfLife = 42;</code>"`;

exports[`should render plaintext without syntax highlighing 1`] = `"<code>export const meaningOfLife = 42;</code>"`;

exports[`should render syntax highlighting 1`] = `"<code class="hljs lang-ts" tabindex="0"><span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> meaningOfLife = <span class="hljs-number">42</span>;</code>"`;

exports[`should render vue as html 1`] = `
"<code class="hljs lang-html" tabindex="0">
        <span class="hljs-tag">&lt;<span class="hljs-name">template</span>&gt;</span>
            <span class="hljs-tag">&lt;<span class="hljs-name">p</span>&gt;</span>lorem ipsum dolor sit amet<span class="hljs-tag">&lt;/<span class="hljs-name">p</span>&gt;</span>
        <span class="hljs-tag">&lt;/<span class="hljs-name">template</span>&gt;</span>

        <span class="hljs-tag">&lt;<span class="hljs-name">script</span>&gt;</span><span class="language-javascript">
            <span class="hljs-keyword">import</span> { defineComponent } <span class="hljs-keyword">from</span> <span class="hljs-string">&quot;vue&quot;</span>;
            <span class="hljs-keyword">export</span> <span class="hljs-keyword">default</span> <span class="hljs-title function_">defineComponent</span>({
                <span class="hljs-attr">name</span>: <span class="hljs-string">&quot;MyAwesomeComponent&quot;</span>,
            });
        </span><span class="hljs-tag">&lt;/<span class="hljs-name">script</span>&gt;</span>
    </code>"
`;
